pandas/_typing.py

```diff
@@ -1,34 +1,37 @@
 from pathlib import Path
-from typing import IO, AnyStr, TypeVar, Union
+from typing import IO, TYPE_CHECKING, AnyStr, TypeVar, Union
 
 import numpy as np
 
-from pandas._libs import Timestamp
-from pandas._libs.tslibs.period import Period
-from pandas._libs.tslibs.timedeltas import Timedelta
+# To prevent import cycles place any internal imports in the branch below
+# and use a string literal forward reference to it in subsequent types
+# https://mypy.readthedocs.io/en/latest/common_issues.html#import-cycles
+if TYPE_CHECKING:
+    from pandas._libs import Period, Timedelta, Timestamp
+
+    from pandas.core.arrays.base import ExtensionArray
+    from pandas.core.dtypes.dtypes import ExtensionDtype
+    from pandas.core.dtypes.generic import (
+        ABCDataFrame,
+        ABCExtensionArray,
+        ABCIndexClass,
+        ABCSeries,
+        ABCSparseSeries,
+    )
+    from pandas.core.indexes.base import Index
+    from pandas.core.frame import DataFrame
+    from pandas.core.series import Series
+    from pandas.core.sparse.series import SparseSeries
 
-from pandas.core.dtypes.dtypes import ExtensionDtype
-from pandas.core.dtypes.generic import (
-    ABCDataFrame,
-    ABCExtensionArray,
-    ABCIndexClass,
-    ABCSeries,
-    ABCSparseSeries,
-)
 
 AnyArrayLike = TypeVar(
-    "AnyArrayLike",
-    ABCExtensionArray,
-    ABCIndexClass,
-    ABCSeries,
-    ABCSparseSeries,
-    np.ndarray,
+    "AnyArrayLike", "ExtensionArray", "Index", "Series", "SparseSeries", np.ndarray
 )
-ArrayLike = TypeVar("ArrayLike", ABCExtensionArray, np.ndarray)
-DatetimeLikeScalar = TypeVar("DatetimeLikeScalar", Period, Timestamp, Timedelta)
-Dtype = Union[str, np.dtype, ExtensionDtype]
+ArrayLike = TypeVar("ArrayLike", "ExtensionArray", np.ndarray)
+DatetimeLikeScalar = TypeVar("DatetimeLikeScalar", "Period", "Timestamp", "Timedelta")
+Dtype = Union[str, np.dtype, "ExtensionDtype"]
 FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
 
-FrameOrSeries = TypeVar("FrameOrSeries", ABCSeries, ABCDataFrame)
+FrameOrSeries = TypeVar("FrameOrSeries", "Series", "DataFrame")
```

Comment:
```suggestion
FrameOrSeries = Union["Series", "DataFrame"]
```

quote from https://mypy.readthedocs.io/en/latest/generics.html...
"User-defined generics are a moderately advanced feature and you can get far without ever using them..."

Reply:
Thanks but this just loosens the type system rather than actually fixing anything. TypeVar is going to be generally more useful for checking functions that can be fully generic in nature.

Might just change the return of this one and see how many others require Union in the future

Reply:
makes sense. Union[Series, DataFrame] _might_ be better written as NDFrame anyway?

Reply:
Also the "user-defined generics" you are referring to are more applicable to containers not TypeVars. Right now we just use a blanket `Series` as a return object, though in the future we could do something like `Series[int]` and `Series[str]`, etc...; the `Series` would be the user-defined generic in that case

The TypeVar in the docs you linked is just a way of parametrizing that user-defined generic, so that a `Series[int]` would have to stay as a `Series[int]` through it's lifecycle; without that parametrization we allow `Series[int]` to become `Series[str]` without any complaints from mypy today

We are probably a ways off of doing user-defined generics but this is great that you looked into it. Certainly open to ideas on that front if you think of a good way to implement as we get more familiar with these annotations

Reply:
> makes sense. Union[Series, DataFrame] _might_ be better written as NDFrame anyway?

Hmm that would work though we don't typically import NDFrame anywhere so I don't think want to start here

Reply:
I would leave as FrameOrSeries as its more descriptive